include/kernels/reduce_core.cuh

```diff
@@ -11,287 +10,253 @@ namespace quda
-    {
-      sum += (ReduceType)a.x * (ReduceType)a.x;
-      sum += (ReduceType)a.y * (ReduceType)a.y;
-      sum += (ReduceType)a.z * (ReduceType)a.z;
-      sum += (ReduceType)a.w * (ReduceType)a.w;
-    }
-
-    template <typename ReduceType> __device__ __host__ void norm2_(ReduceType &sum, const float8 &a)
+    template <typename reduce_t, typename T> __device__ __host__ void norm2_(reduce_t &sum, const typename VectorType<T, 2>::type &a)
     {
-      norm2_(sum, a.x);
-      norm2_(sum, a.y);
+      sum += (reduce_t)a.x * (reduce_t)a.x;
+      sum += (reduce_t)a.y * (reduce_t)a.y;
     }
 
-    template <typename ReduceType, typename Float2, typename FloatN>
-    struct Norm2 : public ReduceFunctor<ReduceType, Float2, FloatN> {
-      Norm2(const Float2 &a, const Float2 &b) { ; }
-      __device__ __host__ void operator()(ReduceType &sum, FloatN &x, FloatN &y, FloatN &z, FloatN &w, FloatN &v)
+    template <typename reduce_t, typename real>
+    struct Norm2 : public ReduceFunctor<reduce_t> {
+      static constexpr write<> write{ };
+      Norm2(const real &a, const real &b) { ; }
+      template <typename T> __device__ __host__ void operator()(reduce_t &sum, T &x, T &y, T &z, T &w, T &v)
       {
-        norm2_<ReduceType>(sum, x);
+#pragma unroll
+        for (int i = 0; i < x.size(); i++) norm2_<reduce_t,real>(sum, x[i]);
       }
-      static int streams() { return 1; } //! total number of input and output streams
-      static int flops() { return 2; }   //! flops per element
+      int streams() const { return 1; } //! total number of input and output streams
+      int flops() const { return 2; }   //! flops per element
     };
 
     /**
        Return the real dot product of x and y
     */
-    template <typename ReduceType> __device__ __host__ void dot_(ReduceType &sum, const double2 &a, const double2 &b)
-    {
-      sum += (ReduceType)a.x * (ReduceType)b.x;
-      sum += (ReduceType)a.y * (ReduceType)b.y;
-    }
-
-    template <typename ReduceType> __device__ __host__ void dot_(ReduceType &sum, const float2 &a, const float2 &b)
-    {
-      sum += (ReduceType)a.x * (ReduceType)b.x;
-      sum += (ReduceType)a.y * (ReduceType)b.y;
-    }
-
-    template <typename ReduceType> __device__ __host__ void dot_(ReduceType &sum, const float4 &a, const float4 &b)
-    {
-      sum += (ReduceType)a.x * (ReduceType)b.x;
-      sum += (ReduceType)a.y * (ReduceType)b.y;
-      sum += (ReduceType)a.z * (ReduceType)b.z;
-      sum += (ReduceType)a.w * (ReduceType)b.w;
-    }
-
-    template <typename ReduceType> __device__ __host__ void dot_(ReduceType &sum, const float8 &a, const float8 &b)
+    template <typename reduce_t, typename T>
+    __device__ __host__ void dot_(reduce_t &sum, const typename VectorType<T, 2>::type &a, const typename VectorType<T, 2>::type &b)
     {
-      dot_(sum, a.x, b.x);
-      dot_(sum, a.y, b.y);
+      sum += (reduce_t)a.x * (reduce_t)b.x;
```

Comment:
Since this is a simple cast from one scalar type to another, the C-style cast and C++ static_cast are identical.  I went with C since it looked more readable.  If there's a good reason to change it, I'm happy to do so.